### PR TITLE
update Option.forall to return true on None (ie, empty)

### PR DESCRIPTION
**Pull request by benfleis**
This fixes behavior noted to be incorrect in #628, and aligns with
typical definition where `all(<empty>) -> true`.

### CODE REVIEW

File: arrow-core/src/main/kotlin/arrow/core/Option.kt [hunk 1]
```diff
@@ -119,18 +119,18 @@ sealed class Option<out A> : OptionKind<A> {
...
-    }
-
     /**
      * Returns true if this option is empty '''or''' the predicate
      * $p returns true when applied to this $option's value.
      *
      * @param p the predicate to test
      */
-    inline fun forall(crossinline p: (A) -> Boolean): Boolean = exists(p)
+    inline fun forall(crossinline p: Predicate<A>): Boolean = fold({ true }, p)
```
Comment (benfleis) [inline arrow-core/src/main/kotlin/arrow/core/Option.kt]: Note that I also put forall underneath exists -- sibling funcs now live next to one another.

File: arrow-data/src/test/kotlin/arrow/data/OptionTest.kt [hunk 2]
```diff
@@ -137,8 +137,14 @@ class OptionTest : UnitSpec() {
 
         "exists" {
             some.exists { it.startsWith('k') } shouldBe true
+            some.exists { it.startsWith('j') } shouldBe false
             none.exists { it.startsWith('k') } shouldBe false
```
Comment (benfleis) [inline arrow-data/src/test/kotlin/arrow/data/OptionTest.kt]: I debated changing the predicate in both None cases to throw a runtime exception, to additionally assert that the predicate itself is never called. Is there already an existing mechanism for this type of assertion in arrow?

Comment (raulraja) [reply arrow-data/src/test/kotlin/arrow/data/OptionTest.kt]: @benfleis if you are referring to do that in the test you can use `kotlin-test` facilities https://github.com/kotlintest/kotlintest/blob/master/doc/reference.md#exceptions- to capture exceptions.

Comment (pakoito) [reply arrow-data/src/test/kotlin/arrow/data/OptionTest.kt]: the exception will propagate and fail the tests. If you'd like to fail the test manually use fail("Message").

Comment (benfleis) [reply arrow-data/src/test/kotlin/arrow/data/OptionTest.kt]: If I update none.exists as follows:

```kotlin
    none.exists { throw NotImplementedError("predicate must not run") } shouldBe false
```

it strengthens the test. I was curious whether that throw above (a la `TODO(...)`) is the standard. Thus, it should not throw -- this variant would simply make it apparent if None's implementation became wrong.

Comment (pakoito) [reply arrow-data/src/test/kotlin/arrow/data/OptionTest.kt]: I'd rather it was a simple `fail()` call as it's part of the framework. Won't block on it tho.

Comment (benfleis) [reply arrow-data/src/test/kotlin/arrow/data/OptionTest.kt]: ah, that's what I was looking for. Have to run, but will try to drop a tiny PR for this later.

**Comment by benfleis** [comment]
@pakoito @raulraja  it appears to work, and one of you has to take it home (after rebase). correct?

**Comment by pakoito** [comment]
@benfleis That's correct. You're on Master now :D Thank you for your contribution. How was it?

**Comment by benfleis** [comment]
feels good, of course.

**Pull request merged**